### DISCUSSION TITLE
How is the name of the cached ZIP file determined?

### DISCUSSION THREAD

**Discussion by tuaris** [composer/composer#9577] [Q&A / Support]
For example, I have this package in my local cache https://repo.packagist.org/packages/aws/aws-sdk-php#3.154.6.  The file is stored as `aws/aws-sdk-php/cce780e22e8a912a59164f230fc2b984b4fd6fed.zip`

Where does `cce780e22e8a912a59164f230fc2b984b4fd6fed` come from?

It's not the commit hash, and it's not an MD5 or SHA1 sum of the archive.  How does composer generate that file name?

**Comment by Seldaek** [comment]
As per https://github.com/composer/composer/blob/2406a094d3b2e5d73ce105764b36898ebd19b69d/src/Composer/Downloader/FileDownloader.php#L106-L115 - it is by default a sha1 of the file URL, but this can be in some cases modified by plugins.

**Reply by tuaris** [reply]
Thanks for the hint.  I suppose my next question now becomes (assuming no plugins are in use), is it possible to determine the URL from only the contents of the `composer.lock` file?  I bit of reverse engineering reveals that composer makes a query to `api.github.com` which then appears redirect to `codeload.github.com`. 

So I think what's going on is it's using the `dist` URL, and that's what redirects to `https://codeload.github.com/aws/aws-sdk-php/legacy.zip/742663a85ec84647f74dea454d2dc45bba180f9d`

Problem is neither of those URL's when hashed come to `cce780e22e8a912a59164f230fc2b984b4fd6fed`.

```
    "packages": [
        {
            "name": "aws/aws-sdk-php",
            "version": "3.171.2",
            "source": {
                "type": "git",
                "url": "https://github.com/aws/aws-sdk-php.git",
                "reference": "742663a85ec84647f74dea454d2dc45bba180f9d"
            },
            "dist": {
                "type": "zip",
                "url": "https://api.github.com/repos/aws/aws-sdk-php/zipball/742663a85ec84647f74dea454d2dc45bba180f9d",
                "reference": "742663a85ec84647f74dea454d2dc45bba180f9d",
                "shasum": ""
            }
```
I am extending an OS's package system to support PHP applications that make use of composer.  The are two important rules.

- During build, no processes may make outbound network connections. 
- Distributions files must be fetched by the build system's own functions. 

Currently I have got it where the build system extracts the list of distribution files for dependencies it needs to download from `composer.lock`.  After downloading them to it's local cache, it moves them into the composer cache directory where it would be arranged as composer would expect it.  The only part left to do is figuring out how to generate the ZIP file name.

**Reply by tuaris** [reply]
hmm... briefly looking at the code it seems like the `processUrl` function is making some changes to the URL grabbed from the `composer.lock` file.  I might need spend a little time tracking down what exactly is changing and see if I might be able to do replicate the same result.
